pkg/querier/querier_test.go

```diff
@@ -1220,6 +1220,101 @@ func TestConfig_ValidateLimits(t *testing.T) {
 	}
 }
 
+func TestClampTime(t *testing.T) {
+	logger := log.NewNopLogger()
+
+	now := time.Now()
+	allowedDeltaT := float64(5 * time.Second.Milliseconds())
+
+	tests := []struct {
+		testName            string
+		originalT           int64
+		referenceT          int64
+		limit               time.Duration
+		clampMax            bool
+		clampForward        bool
+		expectedClampedTime int64
+	}{
+		{
+			// scenario:
+			// * limit set to truncate any query if the query minT is older than 24 hours
+			// * originalT is the original query minT, now - 48 hours
+			// * referenceT is now
+			// * since the query minT can never be older than 24 hours,
+			// the original minT will be clamped forwards in time to now - 24 hours
+			testName:            "clamp minT due to max query lookback",
+			originalT:           now.Add(-48 * time.Hour).UnixMilli(),
+			referenceT:          now.UnixMilli(),
+			limit:               -24 * time.Hour,
+			clampMax:            false,
+			expectedClampedTime: now.Add(-24 * time.Hour).UnixMilli(),
+		},
+		{
+			// scenario:
+			// * limit set to truncate any query if the query maxT more than 1 hour into the future
+			// * originalT is the original query maxT, now + 2 hours
+			// * referenceT is now
+			// * since the query maxT can never be more than 1 hour into the future,
+			// the original maxT will be clamped backwards in time to now + 1 hour
+			testName:            "clamp maxT due to max query into future",
+			originalT:           now.Add(2 * time.Hour).UnixMilli(),
+			referenceT:          now.UnixMilli(),
+			limit:               1 * time.Hour,
+			clampMax:            true,
+			expectedClampedTime: now.Add(1 * time.Hour).UnixMilli(),
+		},
+		{
+			// scenario:
+			// * limit set to truncate the label query range if the query range is longer than 6 hours
+			// * originalT is the original query minT, now - 12 hours
+			// * referenceT is the query maxT, now - 3 hours
+			// * since the entire label query range cannot be longer than 6 hours
+			// the original minT will be clamped forwards in time to query maxT - 6 hours
+			testName:            "clamp minT due to max label query length",
+			originalT:           now.Add(-12 * time.Hour).UnixMilli(),
+			referenceT:          now.Add(-3 * time.Hour).UnixMilli(),
+			limit:               -6 * time.Hour,
+			clampMax:            false,
+			expectedClampedTime: now.Add(-9 * time.Hour).UnixMilli(),
```

Comment:
Is this scenario different in behaviour to the first one? It looks like it's exercising the same logic to me.

Reply:
it is not logically different, but I thought it useful to use a test to show our existing use cases for this clamping